# Binary Operations on NumPy Arrays

**NumPy** is a library for scientific computing in Python. This is one of the libraries that comes in handy when working on machine learning problems. Python is very slow compared to many other languages which is the reason that computation on larger datasets consume a lot of time using Python. **NumPy** is a library written primarily in C to handle large multi-dimensional arrays. It is very fast compared to python as it is written in C and comes very advantageous on larger datasets.

Let us look some binary operations on NumPy arrays which are not very commonly used but can come in need when required. The functions that we are going to look at are:

- bitwise_and
- bitwise_or
- bitwise_xor
- invert
- left_shift
- right_shift

Let's begin by importing Numpy and listing out the functions covered in this notebook.

In [3]:
# Importing NumPy
import numpy as np

In [4]:
# List of functions explained 
function1 = np.bitwise_and
function2 = np.bitwise_or
function3 = np.bitwise_xor
function4 = np.invert
function5 = np.left_shift
function6 = np.right_shift

## Function 1 -  np.bitwise_and

Computes the bitwiseAND of two numpy arrays element-wise. In Python, bitwiseAND is found of using `&`.
<br/>
<br/>
Example:<br/>
Let us see how the bitwiseAND works conceptually. Consider two numbers: <br/>
A = 12 <br/>
B = 5 <br/>

Binary representations of the two numbers are as follows: <br/>
A = 12 => 1100 <br/>
B = 5  => 0101 <br/>

bitwiseAND operator returns a 1 in each bit position where the corresponding bits in both the numbers are 1 else it returns 0
<br/>
A = 12 => 1 1 0 0 <br/>
B = 5  => 0 1 0 1 <br/>
A&B    => 0 1 0 0 => 4 (in decimal representation) <br/>

In [5]:
# Example 1 - working
arr1 = np.array([2, 4, 6, 12, 85, 143, 72])

arr2 = np.array([8, 21, 98, 100, 454, 198, 67])

np.bitwise_and(arr1, arr2)

array([  0,   4,   2,   4,  68, 134,  64])

We can see the working of `bitwise_and` in the above example. The function computes bitwise_and between two NumPy arrays element-wise i.e, 2 AND 8 = 0, 4 AND 21 = 4 and so on.

In [6]:
# Example 2 - working

arr1 = np.array([[1, 2], [3, 4]])

arr2 = np.array([[5, 6], [8, 9]])

np.bitwise_and(arr1,arr2)

array([[1, 2],
       [0, 0]])

We performed `bitwise_and` between two matrices of shape 2X2. We can perform bitwise_and between any two NumPy arrays of same shape.

In [7]:
# Example 3 - breaks
arr1 = np.array([[1, 2], [3, 4]])

arr2 = np.array([[5, 6, 7], [8, 9, 10]])

np.bitwise_and(arr1,arr2)

ValueError: operands could not be broadcast together with shapes (2,2) (2,3) 

In the above example, we did not get the result expected as the shape of arrays `arr1` and `arr2` are different. Therefore, element-wise operations are not possible on this example. We can fix this issue by resizing the arrays to be of equal shape.

## Function 2 - np.bitwise_or

Computes the bitwiseOR of two numpy arrays element-wise. In Python, bitwiseOR is found of using `|`.
<br/>
<br/>
Example:<br/>
Let us see how the bitwiseOR works conceptually. Consider two numbers: <br/>
A = 12 <br/>
B = 5 <br/>

Binary representations of the two numbers are as follows: <br/>
A = 12 => 1100 <br/>
B = 5  => 0101 <br/>

bitwiseOR operator returns a 0 in each bit position where the corresponding bits in both the numbers are 0 else it returns 1
<br/>
A = 12 => 1 1 0 0 <br/>
B = 5  => 0 1 0 1 <br/>
A|B    => 1 1 0 1 => 13 (in decimal representation) <br/>

In [8]:
# Example 1 - working

arr1 = np.array([8, 10, 76, 90])

arr2 = np.array([5, 87, 121, 54])

np.bitwise_or(arr1,arr2)

array([ 13,  95, 125, 126])

We can see the working of `bitwise_or` in the above example. The function computes bitwise_or between two NumPy arrays element-wise i.e, 8 OR 5 = 13, 10 OR 87 = 95 and so on.

In [9]:
# Example 2 - working

arr1 = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])

arr2 = np.array([[14, 15, 16], [17, 18, 19], [20, 21, 22]])

np.bitwise_or(arr1,arr2)

array([[15, 15, 19],
       [21, 23, 23],
       [23, 29, 31]])

We performed `bitwise_or` between two matrices of shape 3X3. We can perform bitwise_or between any two NumPy arrays of same shape.

In [10]:
# Example 3 - breaks

arr1 = np.array([[1, 2], [3, 4]])

arr2 = np.array([[5, 6, 7], [8, 9, 10]])

np.bitwise_or(arr1,arr2)

ValueError: operands could not be broadcast together with shapes (2,2) (2,3) 

In the above example, we did not get the result expected as the shape of arrays `arr1` and `arr2` are different. Therefore, element-wise operations are not possible on this example. We can fix this issue by resizing the arrays to be of equal shape.

## Function 3 - np.bitwise_xor

Computes the bitwiseXOR of two numpy arrays element-wise. In Python, bitwiseXOR is found of using `^`.
<br/>
<br/>
Example:<br/>
Let us see how the bitwiseXOR works conceptually. Consider two numbers: <br/>
A = 12 <br/>
B = 5 <br/>

Binary representations of the two numbers are as follows: <br/>
A = 12 => 1100 <br/>
B = 5  => 0101 <br/>

bitwiseXOR operator returns a 1 in each bit position where the corresponding bits in both the numbers are 1  and 0 each else it returns 0
<br/>
A = 12 => 1 1 0 0 <br/>
B = 5  => 0 1 0 1 <br/>
A^B    => 1 0 0 1 => 9 (in decimal representation) <br/>

In [11]:
# Example 1 - working
arr1 = np.array([12, 76, 97, 198, 210])

arr2 = np.array([84, 32, 64, 900, 523])

np.bitwise_xor(arr1,arr2)

array([ 88, 108,  33, 834, 729])

We can see the working of `bitwise_xor` in the above example. The function computes bitwise_xor between two NumPy arrays element-wise i.e, 12 XOR 84 = 88, 76 XOR 32 = 108 and so on.

In [12]:
# Example 2 - working

arr1 = np.array([[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12], [13, 14, 15, 16]])

arr2 = np.array([[56, 57, 58, 59], [60, 61, 62, 63], [64, 65, 66, 67], [68, 69, 70, 71]])

np.bitwise_xor(arr1,arr2)

array([[57, 59, 57, 63],
       [57, 59, 57, 55],
       [73, 75, 73, 79],
       [73, 75, 73, 87]])

We performed `bitwise_xor` between two matrices of shape 4X4. We can perform bitwise_xor between any two NumPy arrays of same shape.

In [13]:
# Example 3 - breaks

arr1 = np.array([[5, 6], [7, 8]])

arr2 = np.array([[11, 12, 13], [14, 15, 16]])

np.bitwise_or(arr1,arr2)

ValueError: operands could not be broadcast together with shapes (2,2) (2,3) 

In the above example, we did not get the result expected as the shape of arrays `arr1` and `arr2` are different. Therefore, element-wise operations are not possible on this example. We can fix this issue by resizing the arrays to be of equal shape.

## Function 4 - np.invert

Computes the bitwiseNOT of all elements in the numpy array. In Python, bitwiseNOT is found of using `~`.
<br/>
<br/>
Example:<br/>
Let us see how the bitwiseNOT works conceptually. Consider a number: <br/>
A = 5 <br/>

Binary representation of the number is as follows: <br/>
A = 5 => 00000000000000000000000000000101 (32-bit representation)<br/>

bitwiseNOT operator returns a 1 in each bit position where the corresponding bit is 0 and vice-versa.
<br/>
A = 12 => 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 1 <br/>
~A     => 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 0 => -6 (in decimal representation) <br/>

**Note:** *For a signed integer inputs, the two's complement is returned.* `bitwise_not` is an alias for `invert`

In [14]:
# Example 1 - working

arr1 = np.array([12, 124, 9089, 31, 37])

np.invert(arr1)

array([  -13,  -125, -9090,   -32,   -38])

We can see the working of `invert` in the above example. The function computes bitwise_not for all elements in the array i.e, ~12 = -13, ~124 = -125 and so on.

In [15]:
# Example 2 - working
arr1 = np.array([[1,2],[-90, -42]])

np.invert(arr1)

array([[-2, -3],
       [89, 41]])

We performed `bitwise_not` for all elements in the matrix of shape 2X2. We can perform bitwise_not for all NumPy arrays.

In [16]:
# Example 3 - working

x = np.array([True, False])

np.bitwise_not(x)

array([False,  True])

We can see another example using boolean value in the array. There are no counter examples to this `bitwise_not` operation.

## Function 5 - np.left_shift

Computes the left shift operation for all the elements in the array i.e, moves each digit in number's binary representation left. During left shift, the most-significant bit is lost and 0 bit is inserted on the other end. In Python, left-shift is found using `<<`
<br/>
<br/>
Example:<br/>
Let us see how the left-shift works conceptually. Consider a number: <br/>
A = 12 <br/>

Binary representation of the number is as follows: <br/>
A = 12 => 1100 <br/>

Let us say, we want to left-shift this number by 2 bits.

A = 12 => 1 1 0 0 <br/>
A<<2   => 1 1 0 0 0 0 => 48 (in decimal representation) <br/>

In [17]:
# Example 1 - working
arr1 = np.array([42, 87, 93, 76, 543])

np.left_shift(arr1,3)

array([ 336,  696,  744,  608, 4344])

We can see the working of `left_shift` in the above example. The function computes left-shift operation for all elements in the array by 3 bits each i.e, 42<<3 = 336, 87<<3 = 696 and so on.

In [18]:
# Example 2 - working
arr1 = np.array([[1,2,3],[4,5,6],[7,8,9]])

np.left_shift(arr1,2)

array([[ 4,  8, 12],
       [16, 20, 24],
       [28, 32, 36]])

We performed `left-shift` operation by 2 bits for every element in the above matrix of shape 3X3. We can perform left-shift operation for any NumPy array

In [19]:
# Example 3 - working
arr1 = np.array([True,False])

np.left_shift(arr1,3)

array([8, 0])

This operation also works for arrays elements consisting boolean values. It consider `True` as 1 and `False` as 0. In the above example, we used the left-shift operation by 3 bits on the boolean array

**Note:** *We can use the left-shift operation to find out the 2 power N for any positive integer 'N' i.e, (1<<N). This method finds the Nth power of 2 in LogN time*

## Function 6 - np.right_shift

Computes the right shift operation for all the elements in the array i.e, moves each digit in number's binary representation right. During right shift, the least-significant bit is lost and 0 bit is inserted on the other end. In Python, right-shift is found using `>>`
<br/>
<br/>
Example:<br/>
Let us see how the right-shift works conceptually. Consider a number: <br/>
A = 12 <br/>

Binary representation of the number is as follows: <br/>
A = 12 => 1100 <br/>

Let us say, we want to right-shift this number by 2 bits.

A = 12 => 1 1 0 0 <br/>
A>>2   => 0 0 1 1 => 3 (in decimal representation) <br/>

In [21]:
# Example 1 - working

arr1 = np.array([65, 87, 92, 23, 17, 187])

np.right_shift(arr1, 4)

array([ 4,  5,  5,  1,  1, 11])

We can see the working of `right_shift` in the above example. The function computes right-shift operation for all elements in the array by 4 bits each i.e, 65>>4 = 4, 87>>4 = 5 and so on.

In [22]:
# Example 2 - working

arr1 = np.array([[21,22,23,24],[25,26,27,28],[29,30,31,32],[33,34,35,36]])

np.right_shift(arr1,2)

array([[5, 5, 5, 6],
       [6, 6, 6, 7],
       [7, 7, 7, 8],
       [8, 8, 8, 9]])

We performed `right-shift` operation by 2 bits for every element in the above matrix of shape 4X4. We can perform right-shift operation for any NumPy array. The left-shift and right-shift operators have no counter examples where they does not work. These operations work for all types of numpy arrays

In [27]:
# Example 3 - working

arr1 = np.array([True, False])

np.right_shift(arr1,2)

array([0, 0])

This operation also works for arrays elements consisting boolean values. It consider `True` as 1 and `False` as 0. In the above example, we used the right-shift operation by 2 bits on the boolean array

## Conclusion

In the above notebook, we have discussed the use of some numpy arrays operations that we do not frequently use but may require to use them at some point of time. 

## Reference Links
Links to reference and other interesting articles on NumPy:
* Numpy official tutorial : https://numpy.org/doc/stable/user/quickstart.html
* Numpy binary operations: https://numpy.org/doc/stable/reference/routines.bitwise.html#elementwise-bit-operations